ui/src/shared/apis/singleQuery.ts

```diff
@@ -0,0 +1,91 @@
+import {File, CancelBox} from 'src/types'
+import {runQuery} from 'src/shared/apis/query'
+
+/*\
+
+    Declare this at the module level to create a CancelBox mutex
+    for runQuery. this will allow only one query to be run at a time, and
+    pass its results to any calls made to it while waiting for the api.
+    usage:
+      const mutex = RunQueryPromiseMutex()
+
+      function execute(query, state) {
+        // return runQuery(orgID, query, extern, abortController)
+        mutex.run(orgID, query, extern)
+      }
+
+\*/
+export function RunQueryPromiseMutex<T>() {
+  const cached = []
+  let processing = false
+
+  const ret = {
+    run: (orgID: string, query: string, extern?: File): CancelBox<T> => {
+      return {
+        promise: new Promise((resolve, reject) => {
+          if (processing) {
+            cached.push({
+              resolve,
+              reject,
+              cancel: () => {},
+            })
+
+            return
+          }
+
+          processing = true
+          const abortController = new AbortController()
+          cached.push({
+            resolve,
+            reject,
+            cancel: () => {
+              abortController.abort()
+            },
+          })
+
+          runQuery(orgID, query, extern, abortController)
+            .promise.then((...args) => {
+              ret.resolve.apply(ret, args) // eslint-disable-line prefer-spread
+            })
+            .catch((error: Error) => {
+              ret.reject(error)
+            })
+        }),
+        cancel: ret.cancel,
+      }
+    },
+    resolve: (...args) => {
+      let curr
+
+      while (cached.length) {
+        curr = cached.pop()
+        curr.resolve.apply(curr, args) // eslint-disable-line prefer-spread
```

Comment:
curious about the `apply` here rather than just calling `curr.resolve(...args)` since it's using `curr` for context